torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -314,6 +316,326 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+  if (origin_expr != nullptr) {
+    auto inputs = origin_expr->inputs();
+    inputs.assign(expr_inputs.begin(), expr_inputs.end());
+  }
+
+  // Expr* consumer_origin =
+  new UnaryOp(UnaryOpType::Set, consumer, producer);
+
+  // Expr* consumer_uses =
+  for (auto expr : uses_expr)
+    createExprProducer(expr, this, consumer);
+
+  // Before: This TV -> Other TV
+  // After:  This TV -> New TV (After) -> Other TV
+  if (this->hasComputeAt()) {
+    int axis = static_cast<int>(this->getThisComputeAtAxis());
+    this->computeAt(this->getComputeAtView(), axis);
+  }
+
+  // Before: Other Producers -> This TV
+  // After:  Other Producers -> This TV -> New TV (After)
+  // Iterate over origin expression inputs
+  for (Val* v : expr_inputs) {
+    if (v->getValType().value() == ValType::TensorView) {
+      TensorView* origin_input = dynamic_cast<TensorView*>(v);
+      if (origin_input->hasComputeAt() &&
+          origin_input->getComputeAtView()->sameAs(this)) {
+        int axis = static_cast<int>(origin_input->getThisComputeAtAxis());
+        origin_input->computeAt(consumer, axis);
+      }
+    }
+  }
+
+  return consumer;
+}
+
+// Create New Expr given consumer - [output of the expression]
+struct CreateExprConsumer : public OptOutDispatch {
+ public:
+  static void create(Expr* expr, TensorView* current, TensorView* consumer) {
+    CreateExprConsumer dispatch(current, consumer);
+    dispatch.handle(expr);
+  }
+
+ private:
+  explicit CreateExprConsumer(TensorView* current, TensorView* consumer)
+      : current_(current), consumer_(consumer) {}
+
+  void handle(Expr* expr) final {
+    OptOutDispatch::handle(expr);
+  }
+
+  void handle(UnaryOp* unary_expr) final {
+    new UnaryOp(unary_expr->getUnaryOpType(), consumer_, unary_expr->in());
+  }
+
+  void handle(BinaryOp* binary_expr) final {
+    new BinaryOp(
+        binary_expr->getBinaryOpType(),
+        consumer_,
+        binary_expr->lhs(),
+        binary_expr->rhs());
+  }
+
+  void handle(TernaryOp* ternary_expr) final {
+    new TernaryOp(
+        ternary_expr->getTernaryOpType(),
+        consumer_,
+        ternary_expr->in1(),
+        ternary_expr->in2(),
+        ternary_expr->in3());
+  }
+
+  void handle(ReductionOp* reduction_expr) final {
+    if(current_->hasComputeAt())
+      TransformReplay::replayPasC(consumer_, current_, current_->getThisComputeAtAxis());
+    new ReductionOp(
+        reduction_expr->getReductionOpType(),
+        reduction_expr->init(),
+        consumer_,
+        reduction_expr->in());
+  }
+
+  void handle(BroadcastOp* broadcast_expr) final {
+    if(current_->hasComputeAt())
+      TransformReplay::replayPasC(consumer_, current_, current_->getThisComputeAtAxis());
+    new BroadcastOp(consumer_, broadcast_expr->in());
+  }
+
+ private:
+  TensorView* current_;
+  TensorView* consumer_;
+};
+
+// Create New Expr given producer - [an input for the expression]
+struct CreateExprProducer : public OptOutDispatch {
```

Comment:
This should be OptInDispatch, so if it's passed an expr it doesn't recognize it errors instead of doing nothing.